src/TypeInference/Constraints.hs

```diff
@@ -45,6 +45,50 @@ data Constraint a where
   TypeClassNeg :: a -> ClassName -> Typ Neg -> Constraint a
     deriving (Eq, Ord, Functor)
 
+-- | Witnesses generated while solving (sub-)constraints.
+-- Provide a path from the subtype to the supertype and thus
+-- a coercion function.
+data SubtypeWitness
+  = SynL RnTypeName SubtypeWitness
+  -- ^ Witness for type synonym as subtype with substituted subconstraint.
+  | SynR RnTypeName SubtypeWitness
+  -- ^ Witness for type synonym as supertype with substituted subconstraint.
+  | FromTop (Typ Pos)
+  -- ^ Witness for the type being a subtype of /Top/.
+  | ToBot (Typ Neg)
+  -- ^ Witness for the type being a supertype of /Bot/.
+  | Meet SubtypeWitness SubtypeWitness
```

Comment:
Wahrscheinlich besser das in "Union" und "Inter" umzubenennen, damit wir innerhalb der Codebase konsistent sind.